pkg/trie/triedb/node.go

```diff
@@ -168,8 +204,200 @@ func newNodeFromEncoded(nodeHash common.Hash, data []byte, storage NodeStorage)
 			children[i] = child
 		}
 
-		return Branch{partialKey: key, children: children, value: NewFromEncoded(value)}, nil
+		return Branch{partialKey: key, children: children, value: NewValueFromEncoded(encoded.PartialKey, value)}, nil
 	default:
 		panic("unreachable")
 	}
 }
+
+type NodeToEncode interface {
+	isNodeToEncode()
+}
+
+type (
+	NewNodeToEncode struct {
+		partialKey []byte
+		value      []byte
+	}
+	TrieNodeToEncode struct {
+		child NodeHandle
+	}
+)
+
+func (NewNodeToEncode) isNodeToEncode()  {}
+func (TrieNodeToEncode) isNodeToEncode() {}
+
+type ChildReference interface {
+	getNodeData() []byte
+}
+
+type (
+	HashChildReference struct {
+		hash common.Hash
+	}
+	InlineChildReference struct {
+		encodedNode []byte
+	}
+)
+
+func (h HashChildReference) getNodeData() []byte {
+	return h.hash.ToBytes()
+}
+func (i InlineChildReference) getNodeData() []byte {
+	return i.encodedNode
+}
+
+type childFunc = func(node NodeToEncode, partialKey []byte, childIndex *byte) (ChildReference, error)
+
+const emptyTrieBytes = byte(0)
+
+// TODO: move this to codec package
```

Comment:
I would rather recommend creating an issue for this right away, there is nothing more permanent as something expected to be temporal

Reply:
I've been checking and don't make sense to move this code, so I'll remove the TODO